internal/message/amqp/publisher.go

```diff
@@ -0,0 +1,131 @@
+package amqp
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"github.com/streadway/amqp"
+)
+
+const (
+	connectionRetryLimit = 5
+
+	queueDurable    = true
+	queueAutoDelete = false
+	queueExclusive  = false
+	queueNoWait     = false
+
+	amqpExchange     = ""
+	publishMandatory = true
+	publishImmediate = false
+)
+
+var (
+	connectionRetryDelay = 5 * time.Second
+	queueArgs            = amqp.Table{
+		"x-single-active-consumer": true,
+	}
+)
+
+type publisher struct {
+	log       logr.Logger
+	uri       string
+	queueName string
+
+	conn    Connection
+	channel Channel
+	err     chan error
+}
+
+func NewPublisher(uri, name string, logger logr.Logger) (*publisher, error) {
+	p := &publisher{
+		log:       logger.WithName("MessagePublisher"),
+		uri:       uri,
+		queueName: name,
+		err:       make(chan error),
+	}
+
+	if err := p.connect(); err != nil {
+		return nil, err
+	}
+	return p, nil
```

Comment:
You had previously mentioned waiting until `Push` to connect at all? Would this crashloop the pod instead (purposefully presumably)?

Reply:
Sorry, I probably misstated the design earlier but here it is:

1. If you try to create a publisher against an endpoint that doesn't exist, then you get an immediate failure. This makes sense to me from a "fail fast" approach since we can then immediately see that something is up with our AMQP broker.
2. If the AMQP broker is living a happy life, then the connection never has to be re-established. If it dies, then we register the err so that the publisher can attempt a reconnect the next time it needs to use the connection.

Reply:
:+1: